README.md
Evented communication with RFXtrx433.
=====================================

How to Use
==========

rfxcom depends on the serialport module.

To install
----------

<pre>
  npm install rfxcom
</pre>

The only dependency is serialport 1.4.10+.

To Use
------

<pre>
var rfxcom = require('rfxcom'),
    pg = require('pg').native,
    conString = "pg://user:password@localhost/user",
    client = new pg.Client(conString);

var rfxtrx = new rfxcom.RfxCom("/dev/ttyUSB0", {debug: true});

/*
 * This reports security updates from X10 security devices.
 */
rfxtrx.on("security1", function (evt) {
  if (evt.deviceStatus === rfxcom.security.MOTION) {
    console.log("Device %s %s detected motion.", evt.subtype, evt.id);
  } else if (evt.deviceStatus === rfxcom.security.NOMOTION) {
    console.log("Device %s %s reported motion stopped.", evt.subtype, evt.id);
  }
});

rfxtrx.on("elec2", function (evt) {
  // Requires a PostgreSQL table
  // CREATE TABLE energy (recorded_time timestamp DEFAULT NOW(),
  //                      device_id VARCHAR, current_watts FLOAT)
  client.query("INSERT INTO energy(device_id, current_watts) values($1, $2)",
                [evt.id, evt.currentWatts]);
});

rfxtrx.initialise(function () {
    console.log("Device initialised");
});
</pre>

Sending Commands
===
Prototype objects are provided for some of the most useful protocols (see the RFXCOM manual for details):

* Lighting1
* Lighting2
* Lighting3
* Lighting4
* Lighting5
* Lighting6
* Curtain1

Each prototype has a constructor, most of which must be called with the required subtype as a second parameter. The subtypes
are exported from `index.js` and can be accessed as shown in the examples below. Each prototype has functions to send the appropriate
commands.

LightwaveRf
-----------
LightwaveRf devices use the specialised Lighting5 prototype, which itself uses an RfxCom object.

<pre>
var rfxcom = require('rfxcom');

var rfxtrx = new rfxcom.RfxCom("/dev/ttyUSB0", {debug: true}),
    lightwaverf = new rfxcom.Lighting5(rfxtrx, rfxcom.lighting5.LIGHTWAVERF);

rfxtrx.initialise(function () {
  console.log("Device initialised");
  lightwaverf.switchOn("0xF09AC8/1", {mood: 0x03});
  lightwaverf.switchOn("0xF09AC8/2", {level: 0x10});
});
</pre>

I've tested it with both LightwaveRf lights, and the relay switch.

LightwaveRf lights get their identity from the remote used to pair, if you don't
have a remote, or if you want to specify the address manually, you can pair the
device by putting the device into pairing mode and turning on a chosen device id, for example lightwaverf.switchOn("0xFFFFFF/1").

The device ids don't have to be unique, but it's advisable.

HomeEasy (EU)
-------------
HomeEasy devices use the specialised Lighting2 prototype, which itself uses an RfxCom object. There are two types of
HomeEasy: the ones marketed in UK are of type 'AC', while those in the Netherlands and elsewhere are of type 'HOMEEASY_EU'.

<pre>
    var rfxtrx = new rfxcom.RfxCom("/dev/ttyUSB0", {debug: true}),
        lighting2 = new rfxcom.Lighting2(rfxtrx, rfxcom.lighting2.HOMEEASY_EU);

    lighting2.switchOn("0xF09AC8AA/1");
    lighting2.switchOff("0xF09AC8AA/1");
</pre>

The lighting2 message controls one of three subtypes, you need to specify the
subtype to the constructor, the options are in rfxcom.lighting2.


RfxCom system events
====================

System events are used to track conection and disconnection of the RFXtrx433, and to provide
low-level access to received data.

"connecting"
------------
Emitted when the RFXcom has successfully opened the serial port.

"connectfailed"
------------
Emitted if the RFXcom was unable to open the serial port.

"ready"
-------
Emitted when the RFXtrx433 is ready to accept data (after a delay to prevent it from entering the bootloader).

"disconnect"
------------
Emitted if the RFXtrx433 has been disconnected from the USB port

"response"
----------
Emitted when a response message is received from the RFXtrx 433, sends the
status (from the RFXtrx433) and the sequence number of the message the response
is for.

"status"
--------
Emitted when a "status" message is received from the RFXtrx 433.

"end"
--------
Emitted when the serial port "ends".

"receive"
---------
Emitted when any message is received from the RFXtrx 433, sends the raw bytes that were received.

RfxCom received data events
===========================

The events are mostly named from the message identifiers used in the RFXtrx documentation. Not all protocols
can be received (some are transmit-only), and a protocol must be enabled to be received. This can be done using RFXmngr.exe,
or the `enable()` function of the rfxcom object.

"elec1-5"
-------
Emitted when data is received from electricity power & energy monitoring devices
such as the Owl CM119/CM160.

"rfxmeter"
---
Emitted when data is received from an RFXmeter remote meter pulse counter

"rfxsensor"
---
Emitted when data is received from an RFXsensor remote temperature/voltage sensor

"chime"
---
Emitted when data is received from Byron wireless doorbells

"security1"
-----------
Emitted when an X10 security device reports a status change.

"lighting1"
-----------
Emitted when a message is received from X10, ARC, Energenie or similar lighting remote control devices.

"lighting2"
-----------
Emitted when a message is received from AC/HomeEasy type remote control devices.

"lighting4"
-----------
Emitted when a message is received from devices using the PT2262 family chipset.

"lighting5"
-----------
Emitted when a message is received from LightwaveRF/Siemens type remote control devices.

"lighting6"
-----------
Emitted when a message is received from Blyss lighting remote control devices.

"humidity"
---
Emitted when data is received from a LaCrosse humidity sensor.

"th1-9"
-------
Emitted when a message is received from Oregon Scientific
Temperature/Humidity sensors.

"temp1-10"
---------
Emitted when a message is received from an Oregon Scientific temperature.
sensor.

"thb1-2"
---
Emitted when a message is received from Oregon Scientific
temperature, humidity and barometric pressure combined sensors.

"rain"
---
Emitted when a message is received from an Oregon Scientific, LaCrosse,
or similar rain gauge.

"temprain"
----------
Emitted when a message is received from an Alecto combined temperature & rain gauge.

"wind"
---
Emitted when a message is received from an Oregon Scientific or similar wind speed/direction sensor.

"uv"
---
Emitted when a message is received from an Oregon Scientific or similar UV index sensor.
"weight"
---
Emitted when a message is received from an Oregon Scientific weighing scale.

Connecting and disconnecting
===
The function `rfxtrx.initialise()` will attempt to connect to the RFXtrx433 hardware. If this succeeds, a 'connecting' event
is emitted, followed about 5.5 seconds later by a 'ready' event. If the device is not present (wrong device path, or device
not plugged in) a 'connectfailed' event is emitted. If the the hardware is subsequently unplugged, a 'disconnect' event
is emitted (this can happen before either the 'connecting' or 'ready' events are emitted).

If either the connection fails or the RFXtrx433 is unplugged, a subsequent call to `initialise()` will attempt to reconnect.
The 'disconnect'/'connectfailed' handler may make repeated attempts to reconnect,
but <em>must</em> allow an interval of at least `rfxcom.initialiseWaitTime` milliseconds between each attempt. While
disconnected, any data sent by a call to a command object is silently discarded, however the various received data event
handlers are preserved.


changelog.md
Version 0.7.8
-------------

- Fix issue parsing the RSSI data from Lighting2 devices contributed by
  @maxwellhadley

Version 0.7.7
-------------

- Change tbX messages to tbhX messages and fix typo in battery level
  determination.

Version 0.7.6
-------------

- Handle the rfxmeter and tempbaro messages.

Version 0.7.5
-------------

- Implement support for rfxmeter and tempbaro sensors from bwired-nl.

Version 0.7.4
-------------

- Support for serialport 1.3.0 which requires a callback for flush this fixes
  bigkevmcd/node-rfxcom#26.

Version 0.7.3
-------------

- Fix bug in emitting "received" events.

Version 0.7.2
-------------

- Added additional event "received" with the raw data that was received from the
  device contributed by @iangregory.
- Added save() method to the rfxcom object, allowing saving the configured
  protocols to the NVRAM of the rfxtrx433, the SDK cautions against using this
  too often as there is a limited number of writes (10K).

Version 0.7.1
-------------

- Add additional protocol defines for use when configuring the device,
  contributed by @njh.

Version 0.7.0
-------------

- Changes the format of the lighting2 event, the subtype will correspond to the
  values in rfxcom.lighting2.

Version 0.6.1
-------------

- Send a "level" command when we are trying to set the level.

Version 0.6.0
-------------

- Bumped serialport version to be compatible with Node 0.10.x.

Version 0.5.1
-------------

- Removed dependency on underscore.js.

Version 0.5.0
-------------

- Split the various classes out into separate files.
- Split the tests appropriately.
- Rename the LightwaveRf prototype to Lighting5. *backwards incompatible*
- Made the new Lighting5 prototype accept different types.
- Removed RfxCom.lightOn and RfxCom.lightOff because these were LightwaveRf
  specific.

Version 0.4.4
-------------

- Added support for lighting1 received messages contributed by stianeikeland.

Version 0.4.3
-------------

- Added Lighting2 class, which can control devices controlled by the lighting2
  message.

Version 0.4.2
-------------

- Fix the tamper detection for security1 devices to report status and tamper
  separately.

Version 0.4.1
-------------

- Added support for lighting2 received messages.
- Renamed signalStrength to rssi in keeping with the RFX TRX SDK document.

Version 0.4.0
-------------

- Fixed initialise to also run .open() which means you can bootstrap the device
  with one line.

Version 0.3.1
-------------

- Added .initialise to the RfxCom prototype, which handles bootstrapping the
  device.

Version 0.3.0
-------------

- Change the LightwaveRf device identifier format to 0x010203/1 which means that
  it is possible to define device as strings e.g. FRONT_LIGHT = "0x010203/1".

Version 0.2.3
-------------

- Add support for TH1-9 temperature and humidity sensors.

Version 0.2.2
-------------

- Add support for TEMP1-9 type sensors.

Version 0.2.1
-------------

- Fix bug in handler lookup and improve testing of data handler.

Version 0.2.0
-------------

- Update security1Handler event to make the types constants, and better split
  the battery-level and signal strength elements.

Version 0.1.1
-------------

- Updated the Readme.md to illustrate more use.
- Renamed the current_watts and total_watts to currentWatts and totalWatts.

Version 0.1.0
-------------
- *API CHANGE* emitted events are now a JavaScript object.
- New constants for security status messages e.g. security.MOTION.
- Refactored sending messages to the serialport through a sendMessage - still
  need to update the LightwaveRf prototype.
- Improved the test coverage.

Version 0.0.3
-------------
- Added LightwaveRf prototype for specialisation of LightwaveRf control.
  - This adds support for dimming and moodlighting control with LightwaveRF.
- Tidied up JavaScript formatting.
- Added start of rfxcom.security constants.
- Support for configuring the interface protocols.


lib/rfxcom.js
var serialport = require("serialport"),
    rfxcom = require("./index"),
    EventEmitter = require("events").EventEmitter,
    util = require("util");

function RfxCom(device, options) {
    var self = this;

    self.options = options || {};
    self.handlers = {
        0x01: "statusHandler",
        0x02: "messageHandler",
        0x10: "lighting1Handler",
        0x11: "lighting2Handler",
        0x14: "lighting5Handler",
        0x5a: "elec23Handler",
        0x20: "security1Handler",
        0x50: "temp19Handler",
        0x52: "temphumidity19Handler",
        0x54: "tempbaro12Handler",
        0x5D: "weightHandler",
        0x70: "rfxsensorHandler",
        0x71: "rfxmeterHandler"
    };

    // Running counter for command numbers.
    self._cmd = 0;

    // Allow for faking out the SerialPort
    if (typeof self.options.port !== "undefined") {
        self.serialport = options.port;
    }

    // Store the device to use
    self.device = device;

    // This is a buffering parser which accumulates bytes until it receives the
    // number of bytes specified in the first byte of the message.
    // It relies on a flushed buffer, to ensure the first byte corresponds to the
    // size of the first message.
    // The 'data' message emitted has all the bytes from the message.
    self.rfxtrxParser = function() {
        var data = [],
            requiredBytes = null;
        return function(emitter, buffer) {
            // Collect data
            // If we have no required bytes and the number of expected bytes !=
            // 0 then we can store the bytes.
            if (requiredBytes === null && buffer[0] != 00) {
                requiredBytes = buffer[0] + 1;
            }
            if (requiredBytes != null) {
              data.push.apply(data, buffer);
              if (data.length >= requiredBytes) {
                emitter.emit("data", data.slice(0, requiredBytes));
                data = data.slice(requiredBytes);
                requiredBytes = null
              }
            }
        };
    };
}
util.inherits(RfxCom, EventEmitter);

RfxCom.prototype.open = function() {
    var self = this;

    if (typeof self.serialport === "undefined") {
        self.serialport = new serialport.SerialPort(self.device, {
            baudrate: 38400,
            parser: self.rfxtrxParser()
        });
    }
    self.serialport.on("open", function() {
        self.emit("ready");
    });

    // Add data read event listener
    self.serialport.on("data", function(data) {
        if (self.options.debug) {
            console.log("Received: %s", self.dumpHex(data));
        }
        self.emit("receive", data);

        var length = data.shift(),
            packetType = data.shift(),
            handler = self.handlers[packetType];

        if (typeof handler !== "undefined") {
            self[handler](data);
        } else {
            if (self.options.debug) {
                console.log("Unhandled packet type = %s", self.dumpHex([packetType]));
            }
        }
    });

    self.serialport.on("error", function(msg) {
        console.log("Error %s", msg);
    });

    self.serialport.on("end", function() {
        console.log("Received 'end'");
    });

    self.serialport.on("drain", function() {
        console.log("Received 'drain'");
    });
};

RfxCom.prototype.messageHandler = function(data) {
    var self = this,
        subtype = data[0],
        seqnbr = data[1],
        responses = {
            0: "ACK - transmit OK",
            1: "ACK - transmit delayed",
            2: "NAK - transmitter did not lock onto frequency",
            3: "NAK - AC address not allowed"
        },
        message = data[2];
    self.emit("response", responses[message], seqnbr);
};

RfxCom.prototype.initialise = function(callback) {
    var self = this;

    self.on("ready", function() {
        self.reset(function() {
            self.delay(500);
            self.flush(function() {
                self.getStatus(callback);
            });
        });
    });
    self.open();
};


/**
 *
 * Called by the data event handler when data arrives from the device with
 * information about its settings.
 *
 */
RfxCom.prototype.statusHandler = function(data) {
    var self = this,
        receiverTypes = {
            0x50: "310MHz",
            0x51: "315MHz",
            0x52: "433.92MHz receiver only",
            0x53: "433.92MHz transceiver",
            0x55: "868.00MHz",
            0x56: "868.00MHz FSK",
            0x57: "868.30MHz",
            0x58: "868.30MHz FSK",
            0x59: "868.35MHz",
            0x5A: "868.35MHz FSK",
            0x5B: "868.95MHz"
        },
        subtype = data[0],
        seqnbr = data[1],
        cmnd = data[2],
        msg = data.slice(2, 13),
        receiverType = receiverTypes[msg[1]],
        firmwareVersion = msg[2],
        protocols = [];

    // Check which protocols are enabled
    for (var key in rfxcom.protocols) {
        var value = rfxcom.protocols[key];
        if (msg[value.msg] & value.bit) {
            protocols.push(key);
        }
    }

    self.emit("status", {
        subtype: subtype,
        seqnbr: seqnbr,
        cmnd: cmnd,
        receiverType: receiverType,
        firmwareVersion: firmwareVersion,
        enabledProtocols: protocols
    });
};

/**
 *
 * Fetches a "command number" sequence number for identifying requests sent to
 * the device.
 *
 */
RfxCom.prototype.getCmdNumber = function() {
    var self = this;

    if (self._cmd > 256) {
        self._cmd = 0;
    }
    return self._cmd++;
};

/**
 *
 * Internal function for sending messages to the device.
 *
 */
RfxCom.prototype.sendMessage = function(type, subtype, cmd, extra, callback) {
    var self = this,
        cmdId = this.getCmdNumber(),
        byteCount = extra.length + 4,
        buffer = [byteCount, type, subtype, cmdId, cmd];

    buffer = buffer.concat(extra);
    self.serialport.write(buffer, function(err, response) {
        if (self.options.debug) {
            console.log("Sent    : %s", self.dumpHex(buffer));
        }

        if (callback) {
            return callback(err, response, cmdId);
        }
    });

    return cmdId;
};


/**
 *
 * Writes the reset sequence to the RFxtrx433.
 *
 */
RfxCom.prototype.reset = function(callback) {
    var self = this;
    return self.sendMessage(0, 0, 0, [0, 0, 0, 0, 0, 0, 0, 0, 0], callback);
};

/**
 *
 * Calls flush on the underlying SerialPort.
 *
 */
RfxCom.prototype.flush = function(callback) {
    var self = this;
    self.serialport.flush(callback);
};


/*
 * Sends the getstatus bytes to the interface.
 */
RfxCom.prototype.getStatus = function(callback) {
    var self = this;
    return self.sendMessage(0, 0, 2, [0, 0, 0, 0, 0, 0, 0, 0, 0], callback);
};

/*
 * Enables reception of different protocols.
 */
RfxCom.prototype.enable = function(protocols, callback) {
    var self = this,
        msg = [0x53, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00];

    if (!(protocols instanceof Array)) {
        protocols = [protocols];
    }

    protocols.forEach(function(protocol) {
        if (typeof msg[protocol.msg] === "undefined") {
            msg[protocol.msg - 1] = protocol.bit;
        } else {
            msg[protocol.msg - 1] |= protocol.bit;
        }
    });

    return self.sendMessage(0, 0, 0x03, msg, callback);
};


/*
 * Save the enabled protocols of the receiver/transceiver in non-volatile memory
 *
 * Important: Do not send the save command very often because there is a 
 * maximum of 10,000 write cycles to non-volatile memory!
 */
RfxCom.prototype.save = function(callback) {
    var self = this,
        msg = [0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00];

    return self.sendMessage(0, 0, 0x06, msg, callback);
};


/**
 *
 * Wait for at least the specified number of milliseconds.
 *
 * This may not wait precisely for that number of milliseconds due to the
 * vagaries of system scheduling, but no fewer than the specified ms.
 *
 */
RfxCom.prototype.delay = function(ms) {
    ms += +new Date();
    while (+new Date() < ms) {
        // Do nothing...
    }
};

RfxCom.prototype.dumpHex = function(buffer, prefix) {
    prefix = prefix || "";

    function dec2hex(value) {
        var hexDigits = "0123456789ABCDEF";
        return prefix + (hexDigits[value >> 4] + hexDigits[value & 15]);
    }
    return buffer.map(dec2hex);
};

/**
 *
 * Converts an array of 4 bytes to a 32bit integer.
 *
 */
RfxCom.prototype.bytesToUint32 = function(bytes) {
    return (bytes[0] * Math.pow(2, 24)) + (bytes[1] << 16) + (bytes[2] << 8) + bytes[3];
};

/**
 *
 * Converts an array of 6 bytes to a 48bit integer.
 *
 */
RfxCom.prototype.bytesToUint48 = function(bytes) {
    return (bytes[0] * Math.pow(2, 40)) + (bytes[1] * Math.pow(2, 32)) + (bytes[2] * Math.pow(2, 24)) + (bytes[3] << 16) + (bytes[4] << 8) + bytes[4];
};

/**
 *
 * Converts a string to an array of Byte values.
 *
 * e.g. stringToBytes("202020") == [32, 32, 32]
 *
 */
RfxCom.prototype.stringToBytes = function(bytes) {
    var result = [],
        byteString,
        value,
        i;
    for (i = 0; i < bytes.length; i += 2) {
        byteString = bytes.substr(i, 2);
        if (byteString !== "0x") {
            value = parseInt(bytes.substr(i, 2), 16);
            if (typeof value !== "undefined") {
                result.push(value);
            }
        }
    }
    return result;
};

/**
 *
 * Called by the data event handler when data arrives from an OWL CM119/CM160
 * power measurement device.
 *
 * Calculates the current usage and total usage from the bytes sent, and emits
 * an "elec2" and "elec3" messages for handling.
 *
 */
RfxCom.prototype.elec23Handler = function(data) {
    var self = this,
        TOTAL_DIVISOR = 223.666,
        subtype = data[0],
        seqnbr = data[1],
        id = "0x" + self.dumpHex(data.slice(2, 4), false).join(""),
        count = data[4],
        instant = data.slice(5, 9),
        total = data.slice(9, 15),
        currentWatts = self.bytesToUint32(instant),
        totalWatts = self.bytesToUint48(total) / TOTAL_DIVISOR,
        roundedTotal = Math.round(totalWatts * Math.pow(10, 2)) / Math.pow(10, 2),
        evt = {
            subtype: subtype,
            id: id,
            count: count,
            currentWatts: currentWatts,
            totalWatts: roundedTotal
        };

    self.emit("elec" + (subtype + 1), evt);
};

/**
 *
 * Called by the data event handler when data arrives from various security
 * devices.
 *
 */
RfxCom.prototype.security1Handler = function(data) {
    var self = this,
        subtype = data[0],
        seqnbr = data[1],
        id = "0x" + self.dumpHex(data.slice(2, 5), false).join(""),
        deviceStatus = data[5] & ~0x80,
        batterySignalLevel = data[6],
        evt = {
            subtype: subtype,
            id: id,
            deviceStatus: deviceStatus,
            batteryLevel: batterySignalLevel & 0x0f,
            rssi: batterySignalLevel >> 4,
            tampered: data[5] & 0x80
        };

    self.emit("security1", evt);
};

/**
 *
 * Called by the data event handler when data arrives from temp1-9
 * devices.
 *
 */
RfxCom.prototype.temp19Handler = function(data) {
    var self = this,
        subtype = data[0],
        seqnbr = data[1],
        id = "0x" + self.dumpHex(data.slice(2, 4), false).join(""),
        temperature = ((data[4] & 0x7f) * 256 + data[5]) / 10,
        signbit = data[4] & 0x80,
        batterySignalLevel = data[6],
        evt = {
            subtype: subtype,
            id: id,
            seqnbr: seqnbr,
            temperature: temperature * (signbit ? -1 : 1),
            batteryLevel: batterySignalLevel & 0x0f,
            rssi: batterySignalLevel >> 4,
        };
    self.emit("temp" + subtype, evt);
};

/**
 *
 * Called by the data event handler when data arrives from th1-9
 * devices.
 *
 */
RfxCom.prototype.temphumidity19Handler = function(data) {
    var self = this,
        subtype = data[0],
        seqnbr = data[1],
        id = "0x" + self.dumpHex(data.slice(2, 4), false).join(""),
        temperature = ((data[4] & 0x7f) * 256 + data[5]) / 10,
        signbit = data[4] & 0x80,
        humidity = data[6],
        humidityStatus = data[7],
        batterySignalLevel = data[8],
        evt = {
            subtype: subtype,
            id: id,
            seqnbr: seqnbr,
            temperature: temperature * (signbit ? -1 : 1),
            humidity: humidity,
            humidityStatus: humidityStatus,
            batteryLevel: batterySignalLevel & 0x0f,
            rssi: batterySignalLevel >> 4,
        };

    self.emit("th" + subtype, evt);
};

/**
 *
 * Called by the data event handler when data arrives from a Lighting1
 * light control device.
 *
 */
RfxCom.prototype.lighting1Handler = function(data) {
    var self = this,
        subtypes = {
            0: "X10",
            1: "ARC",
            2: "ELRO AB400D",
            3: "Waveman",
            4: "EMW200",
            5: "Impuls",
            6: "RisingSun",
            7: "Philips SBC"
        },
        commands = {
            0: "Off",
            1: "On",
            5: "All Off",
            6: "All On",
            7: "Chime"
        },
        subtype = subtypes[data[0]],
        seqnbr = data[1],
        housecode = String.fromCharCode(data[2]),
        unitcode = data[3],
        command = commands[data[4]],
        rssi = data[5] >> 4,
        id = self.dumpHex(data.slice(2, 4), false).join(""),
        evt;

    evt = {
        id: "0x" + id,
        subtype: subtype,
        seqnbr: seqnbr,
        housecode: housecode,
        unitcode: unitcode,
        command: command,
        rssi: rssi
    };

    self.emit("lighting1", evt);
};

/**
 *
 * Called by the data event handler when data arrives from a LightwaveRF/Siemens
 * light control device.
 *
 */
RfxCom.prototype.lighting2Handler = function(data) {
    var self = this,
        commands = {
            0: "Off",
            1: "On",
            2: "Set Level",
            3: "Group Off",
            4: "Group On",
            5: "Set Group Level"
        },
        subtype = data[0],
        seqnbr = data[1],
        idBytes = data.slice(2, 6),
        unitcode = data[6],
        command = commands[data[7]],
        level = data[8],
        rssi = (data[9] & 0xf0) >> 4,
        evt;

    idBytes[0] &= ~0xfc; // "id1 : 2"
    evt = {
        subtype: subtype,
        seqnbr: seqnbr,
        id: "0x" + self.dumpHex(idBytes, false).join(""),
        unitcode: unitcode,
        command: command,
        level: level,
        rssi: rssi
    };

    self.emit("lighting2", evt);
};

/**
 *
 * Called by the data event handler when data arrives from a LightwaveRF/Siemens
 * light control device.
 *
 */
RfxCom.prototype.lighting5Handler = function(data) {
    var self = this,
        subtypes = {
            0: "LightwaveRF, Siemens",
            1: "EMW100 GAO/Everflourish"
        },
        commands = {
            0: "Off",
            1: "On"
        },
        subtype = subtypes[data[0]],
        seqnbr = data[1],
        id = "0x" + self.dumpHex(data.slice(2, 5), false).join(""),
        unitcode = data[5],
        command = commands[data[6]],
        evt = {
            subtype: subtype,
            id: id,
            unitcode: unitcode,
            command: command,
            seqnbr: seqnbr
        };
    self.emit("lighting5", evt);
};

 /**
   *
   * Called by the data event handler when data arrives from rfxmeter
   * devices.
   *
   */
RfxCom.prototype.rfxmeterHandler = function(data) {
  var self = this,
      id = "0x" + self.dumpHex(data.slice(2, 4), false).join(""),
      counter = self.dumpHex(data.slice(4, 8), false).join(""),
      evt = {
        subtype: data[0],
        id: id,
        seqnbr: data[1],
        counter: parseInt(counter,16),
      };
    self.emit("rfxmeter", evt);
};

/**
  *
  * Called by the data event handler when data arrives from digital
  * scales.
  *
  */
RfxCom.prototype.weightHandler = function(data) {
  var self = this,
      id = "0x" + self.dumpHex(data.slice(2, 4), false).join(""),
      batterySignalLevel = data[6],
      evt = {
        subtype: data[0],
        id: id,
        seqnbr: data[1],
        weight: (data[4] *256 + data[5]) / 10,
        rssi: batterySignalLevel & 0x0f,
        batteryLevel: batterySignalLevel >> 4,
      };
    self.emit("weight" + evt.subtype, evt);
};

/**
  * Called by the data event handler when data arrives from rfxmeter
  * devices.
  *
  */
RfxCom.prototype.rfxsensorHandler = function(data) {
  var self = this,
      subtype = data[0],
      seqnbr = data[1],
      id = "0x" + self.dumpHex([data[2]], false).join(""),
      evt = {
        subtype: subtype,
        id: id,
        seqnbr: seqnbr,
        rssi: data[5] >> 4,
      };

    switch(evt.subtype) {
        case rfxcom.rfxsensor.TEMP:
        var signbit = data[3] & 0x80;
        evt.message =  (((data[3] & 0x7f) * 256 + data[4]) / 100) * (signbit ? -1 : 1);
        break;
        case rfxcom.rfxsensor.VOLTAGE:
        case rfxcom.rfxsensor.AD:
        evt.message = data[3] * 256 + data[4];
        break
    }
    self.emit("rfxsensor", evt);
};

/**
 *
 * Called by the data event handler when data arrives from thb1-thb2
 * devices.
 *
 */
RfxCom.prototype.tempbaro12Handler = function(data) {
   var self = this,
       id = "0x" + self.dumpHex(data.slice(2, 4), false).join(""),
       temperature = ((data[4] & 0x7f) * 256 + data[5]) / 10,
       signbit = data[4] & 0x80,
       barometer = ((data[8] & 0x7f) * 256 + data[9]),
       batterySignalLevel = data[11],
       evt = {
         subtype: data[0],
         id: id,
         seqnbr: data[1],
         temperature: temperature * (signbit ? -1 : 1),
         humidity: data[6],
         humidityStatus: data[7],
         barometer: barometer,
         forecast: data[10],
         batteryLevel: batterySignalLevel & 0x0f,
         rssi: batterySignalLevel >> 4,
      };
  self.emit("thb" + evt.subtype, evt);
};

module.exports = RfxCom;
